src/Elastic.Apm.AspNetCore/Extensions/RequestExtentions.cs

```diff
@@ -16,26 +16,25 @@ public static class HttpRequestExtensions
 		/// <param name="request"></param>
 		/// <param name="logger"></param>
 		/// <returns></returns>
-		public static string ExtractRequestBody(this HttpRequest request, IApmLogger logger)
+		public static async Task<string> ExtractRequestBodyAsync(this HttpRequest request, IApmLogger logger)
 		{
-			string body = null;
+			var body = Apm.Consts.Redacted;
 
 			try
 			{
-				request.EnableRewind();
+				request.EnableBuffering();
```

Comment:
I did a little bit of a reading on this.

> I got an impression from this article that EnableBuffering should be called before any attempt to read the body

Why do you get that impression?  I did not think that by reading that article, nor has the [docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httprequestrewindextensions.enablebuffering?view=aspnetcore-2.1) contain that info. 

Furthermore, we have tests (and I also tried manually) when we read the request body on errors, which means the body is first read by some ASP.NET Core component and then by us. And it works. 

> We should definitely add test for this use case.

So, cases I think are relevant for this are already covered. For example `TestJsonBodyRetrievalOnRequestFailureInAspNetCore` is 1 test where we definitely read the request body very late - basically right before the request ends and I also think that your `Options_changed_after_start` test covers that.

Reply:
I'll leave it for you to decide if it's worth further work - it's quite possible that tests pass because the body size is very small and thus arrives in one TCP packet and/or we don't send request body in multiple chunks (using `Transfer-Encoding: chunked`) or maybe some other reason.
I looked at the source code a little bit
- [`Microsoft.AspNetCore.Http.HttpRequestRewindExtensions.EnableBuffering` just calls `Microsoft.AspNetCore.Http.Internal.BufferingHelper.EnableRewind`](https://github.com/aspnet/AspNetCore/blob/v3.0.0/src/Http/Http/src/Extensions/HttpRequestRewindExtensions.cs#L23)
- [`EnableRewind` wraps `request.Body` into a buffering stream](https://github.com/aspnet/AspNetCore/blob/v3.0.0/src/Http/Http/src/Internal/BufferingHelper.cs#L24) which I don't think can do buffering retroactively

Reply:
I tested this with a very long request body, and it still seems to work. I leave it for now.